Set up REST API

tripu opened the pull request:
This PR sets up a REST API using Express, with 3 methods:

* `version`
* `metadata`
* `validate`

And adds a test file for Mocha with placeholders for API tests.

Because this functionality was moved here from branch `tripu/metadata`, there are a few files in common (`sink.js`, `util.js`, `samples.json`). Those should be OK when we merge the branches.

As the API relies on the new method `Specberus.extractMetadata()`, **do *not* merge this before #299**; it won't work.

**TO-DO:** [Updated 3/May/2016]

* ~~Write the tests. My plan is to launch the Specberus server locally first, then use `superagent` to throw a few requests at it, and check the results.~~ :ballot_box_with_check:
* ~~POST or GET? Method `version` should be GET. I'm not sure about the other two.~~ :ballot_box_with_check: All are GET.
* ~~Refactor if/else block for `api/metadata` and `api/validate`: the signature and checks are almost identical for both.~~ :ballot_box_with_check:

tripu commented:
Fixes #320.

tripu commented:
@deniak, ready for review.
(The 2 errors in the test suite aren't specific about the API; they are caused by #347.)

deniak commented on lib/api.js:
```diff
@@ -0,0 +1,91 @@
+const parseSource = function(query) {
+    const result = {};
+    if (query.url) result.url = query.url;
+    else if (query.source) result.source =uery.source;
```
s/uery/query/

tripu replied:
Fixed, thanks!

deniak commented on README.md:
```diff
@@ -119,6 +121,25 @@ This is an example of the value of `Specberus.meta` after the execution of `Spec
+
+* `<host>/api/version`
+* `<host>/api/metadata?url=http://example.com/doc.html`
+* `<host>/api/validate?file=/home/me/docs/spec.html`
```
That method ('file' QS param) won't work unless the file is already on the server which won't happen. Same thing for 'document'.
Regarding 'source', it seems difficult to pass the whole source as the parameter.
I suggest to handle *only* 'url' for now.

tripu replied:
That caveat about `file` and `document` isn't new. Those two options don't work for users of our public instance of Specberus, but may come handy in other circumstances, eg we use `file` for testing the rules (and I'm using it now to test the REST API too). Or somebody may want to install Specberus locally and pass along `file`s.

@deniak, I'll modify my tests to test both with `file` and `url`, and maybe exclude the other two from the REST API, OK?

deniak replied:
`file` and `document` are definitely useful but so far, we know, it's not working properly and these are new features compared to the current pubrules, so :+1: to disabling these options from the REST API for now.

deniak commented on lib/api.js:
```diff
@@ -0,0 +1,91 @@
+            throw new Error('Cannot retrieve profile “' + query.profile + '”');
+    }
+    if (query.validation) result.validation = query.validation;
+    if (query.processDocument) result.processDocument = query.processDocument;
```
Other options are possible (rectrack, echidna, informative, etc...)

tripu replied:
You are right, I should be accepting the complete set of parameters.

@deniak, I'm now refactoring the parsing of parameters, because otherwise it'd be duplicated in different places. That common logic will be used then by `lib/validator.js` and `lib/api.js` to parse their parameters. Sounds good?

deniak replied:
@tripu WFM. Don't forget to take the profile `auto` into consideration.

tripu commented:
@deniak, I think I addressed all comments, and API tests are now more comprehensive.
Take a look, please.

Since the nsp vulnerability is fixed, and this introduces a new interface for Specberus, when merged I'll bump the minor version and publish, and also update the version of Specberus in Echidna, OK?

deniak commented on lib/api.js:
```diff
@@ -0,0 +1,122 @@
+        catch (err) {
+            return res.status(400).send(err.toString());
+        }
+        if (validate && 'auto' === options.profile) {
```
Trying to validate an url with profile `auto` returns the profile only, eg.
`/api/validate?url=http://www.w3.org/TR/2016/WD-charmod-norm-20160407/&profile=auto` gives `WD`

deniak replied:
Looks good now

tripu commented:
@deniak, now `metadata` and `validate` return a JSON object with errors, warnings, metadata, etc. It's documented [on the README](https://github.com/w3c/specberus/blob/tripu/api/README.md#5-rest-api).

/cc @bert-github, who I think is interested in using Specberus programmatically via the REST API.

deniak commented on lib/api.js:
```diff
@@ -0,0 +1,154 @@
+            );
+        }
+        catch (err) {
+            return res.status(400).send(err.toString());
```
Since the API is expected to return a JSON, it might be better to send a JSON in case of errors too. Same comment for line 145.

deniak commented:
If the API is used on an url that doesn't exist, the exception isn't caught and the page loads indefinitely.

tripu commented:
@deniak, back to you.
